specs/CustomDownload.md

```diff
@@ -0,0 +1,771 @@
+  /// Result file path with file name is too long.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_FILE_NAME_TOO_LONG,
+  /// File is too large for file system.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_FILE_TOO_LARGE,
+  /// File contains a virus.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_FILE_VIRUS_INFECTED,
+  /// File was in use, too many files opened, or out of memory.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_FILE_TRANSIENT_ERROR,
+  /// File blocked by local policy.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_FILE_BLOCKED,
+  /// Security check failed unexpectedly.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_FILE_SECURITY_CHECK_FAILED,
+  /// Seeking past the end of a file in opening a file, as part of resuming an
+  /// interrupted download.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_FILE_TOO_SHORT,
+  /// Partial file did not match the expected hash.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_FILE_HASH_MISMATCH,
+  /// Source and target of download were the same.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_FILE_SAME_AS_SOURCE,
+
+  /// Generic network error.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_NETWORK_FAILED,
+  /// Network operation timed out.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_NETWORK_TIMEOUT,
+  /// Network connection lost.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_NETWORK_DISCONNECTED,
+  /// Server has gone down.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_NETWORK_SERVER_DOWN,
+  /// Network request invalid because original or redirected URI is invalid, has
+  /// an unsupported scheme, or is disallowed by policy.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_NETWORK_INVALID_REQUEST,
+
+  /// Generic server error.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_SERVER_FAILED,
+  /// Server does not support range requests.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_SERVER_NO_RANGE,
+  /// Server does not have the requested data.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_SERVER_BAD_CONTENT,
+  /// Server did not authorize access to resource.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_SERVER_UNAUTHORIZED,
+  /// Server certificate problem.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_SERVER_CERT_PROBLEM,
+  /// Server access forbidden.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_SERVER_FORBIDDEN,
+  /// Unexpected server response. Responding server may not be intended server.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_SERVER_UNREACHABLE,
+  /// Server sent fewer bytes than the content-length header. Content-length
+  /// header may be invalid or connection may have closed. Download is treated
+  /// as complete unless there are strong validators present to interrupt the
+  /// download.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_SERVER_CONTENT_LENGTH_MISMATCH,
+  /// Unexpected cross-origin redirect.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_SERVER_CROSS_ORIGIN_REDIRECT,
+
+  /// User canceled the download.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_USER_CANCELED,
+  /// User shut down the WebView.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_USER_SHUTDOWN,
```

Comment:
How much time does the app have in order to clean up / save state before everything disappears? Does it have to do it all before returning from the StateChanged event handler?

Reply:
That sounds like it should be a more general notification than just if there happens to be a download going on.

Reply:
General notification is not necessary: DOM already has beforeunload event, and app owns close. Downloads may not need this? Will you even receive the event? Might be useful for resuming downloads later after bringing back up but we don't have that feature yet. Should update docs somewhere to note that we don't support that scenario yet.